### PR TITLE
[System tests] Buying deposit with surplus

### PR DESCRIPTION
**Pull request by dimpar**
When there's enough TBTC funds left as surplus to buy out a new deposit
in liquidation, then a new auction should not be opened.

Refs https://github.com/keep-network/coverage-pools/issues/83

### CODE REVIEW

File: test/system/surplus-for-auction.test.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+// about liquidated deposit so it can early close the auction.
+// All the tests below are executed on Hardhat Network with mainnet forking enabled.
+// At the start, the fork is being reset to the specific starting block which
+// determines the initial test state. This test uses the real tBTC token contract
+// and two deposits:
+// https://allthekeeps.com/deposit/0x8495732aecd7f132eaab61f64858ccc73475973f 5 TBTC
+// https://allthekeeps.com/deposit/0xfc9c50fd44879bd7085edd311bc8e2b7d3e41595 1 TBTC
+// which are ready to be liquidated at the starting block. All the bidders are also
+// real accounts with actual TBTC balance. At the end of the scenario, the risk
+// manager should early close the cov pool auction and keep the surplus of TBTC.
```
Comment (michalinacienciala) [inline test/system/surplus-for-auction.test.js:27]: I would consider adding a sentence about the second deposit being bought by the cov pool using the surplus TBTC.

Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Sure 7eeb4c7

File: test/system/surplus-for-auction.test.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      .approveFirstRiskManager(riskManagerV1.address)
+
+    tbtcDeposit = await ethers.getContractAt("IDeposit", depositAddress)
+    tbtcDeposit1 = await ethers.getContractAt("IDeposit", depositAddress1)
+
+    bidder = await impersonateAccount(bidderAddress)
+    bidder1 = await impersonateAccount(bidderAddress1)
+  })
+
+  describe("when bying auction with surplus TBTC funds", () => {
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: 9e1ac79

File: test/system/surplus-for-auction.test.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        gasLimit: 350000,
+      })
+
+      // Risk Manager should not open a new auction for tbtcDeposit1.
+      const auctionAddress = await riskManagerV1.depositToAuction(
+        tbtcDeposit1.address
+      )
+      expect(auctionAddress).to.equal(ZERO_ADDRESS)
+
+      // 1.5 TBTC should be left on Risk Manager after puchasing 0xfc9..
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: 9e1ac79

File: test/system/surplus-for-auction.test.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      // buying 0x849.. deposit outside the coverage pool
+      await tbtcDeposit.connect(bidder1).purchaseSignerBondsAtAuction()
+
+      await riskManagerV1.notifyLiquidated(tbtcDeposit.address)
+
+      expect(await tbtcDeposit.currentState()).to.equal(11) // LIQUIDATED
+    })
+
+    it("should buy deposit 0xfc9.. with surplus TBTC without opening an auction", async () => {
+      expect(await tbtcDeposit1.currentState()).to.equal(10) // IN LIQUIDATION
```
Comment (lukasz-zimnoch) [inline test/system/surplus-for-auction.test.js:156]: Isn't it a good candidate for the `test initial state` section we have in other tests?

**Review by lukasz-zimnoch — CHANGES_REQUESTED** [review]

**Comment by lukasz-zimnoch** [comment]
Please note the changes introduced by #88 before merging. `SignerBondsEscrow` has been renamed.

**Comment by dimpar** [comment]
@lukasz-zimnoch After merging main to this branch I had to refactor these tests. Please find all the changes for your comments above in the current file in this commit: a3840e7

Variable names in other tests were modified because of the common tests setup and your comments above.

File: test/system/surplus-for-auction.test.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+// At the end of the scenario, the risk manager should keep the surplus of TBTC
+// for potential next deposit buy outs.
+
+describeFn("System -- buying a deposit with surplus", () => {
+  const startingBlock = 11536431
+  const tbtcTokenAddress = "0x8daebade922df735c38c80c7ebd708af50815faa"
+  const depositAddress = "0x8495732aecd7f132eaab61f64858ccc73475973f"
+  const depositAddress1 = "0xfc9c50fd44879bd7085edd311bc8e2b7d3e41595"
+  const bidderAddress = "0xa0216ED2202459068a750bDf74063f677613DA34" // 12.36 TBTC
+  const bidderAddress1 = "0xf9e11762d522ea29dd78178c9baf83b7b093aacc" // 10.29 TBTC
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

File: test/system/surplus-for-auction.test.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  let signerBondsSwapStrategy
+  let coveragePool
+  let riskManagerV1
+  let tbtcDeposit
+  let tbtcDeposit1
+
+  let governance
+  let rewardsManager
+  let bidder
+  let bidder1
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

File: test/system/surplus-for-auction.test.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+      const auction = new ethers.Contract(auctionAddress, Auction.abi, bidder)
+      const bidderTake = lotSize.div(2) // 5 / 2 = 2.5 TBTC
+      await tbtcToken.connect(bidder).approve(auction.address, bidderTake)
+
+      // bidder takes a partial offer on a cov pool auction
+      await auction.takeOffer(bidderTake)
+    })
+
+    it("should have TBTC funds for partially selling an auction for deposit 0x849..", async () => {
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

File: test/system/surplus-for-auction.test.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+      const tbtcSurplus = await tbtcToken.balanceOf(riskManagerV1.address)
+      expect(tbtcSurplus).to.be.equal(lotSize.div(2))
+    })
+
+    it("should buy deposit 0x849.. outside the coverage pool", async () => {
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

File: test/system/surplus-for-auction.test.js [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    let surplusTx
+
+    before(async () => {
+      await riskManagerV1.notifyLiquidation(tbtcDeposit.address)
+
+      const auctionAddress = await riskManagerV1.depositToAuction(
+        tbtcDeposit.address
+      )
+
+      const auction = new ethers.Contract(auctionAddress, Auction.abi, bidder)
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

File: test/system/surplus-for-auction.test.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      const bidderTake = lotSize.div(2) // 5 / 2 = 2.5 TBTC
+      await tbtcToken.connect(bidder).approve(auction.address, bidderTake)
+
+      // bidder takes a partial offer on a cov pool auction
+      await auction.takeOffer(bidderTake)
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

File: test/system/surplus-for-auction.test.js [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+    })
+
+    it("should have TBTC funds for partially selling an auction for deposit 0x849..", async () => {
+      const tbtcSurplus = await tbtcToken.balanceOf(riskManagerV1.address)
+      expect(tbtcSurplus).to.be.equal(lotSize.div(2))
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js:128]: Done

File: test/system/surplus-for-auction.test.js [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+      // Risk Manager should not open a new auction for tbtcDeposit1.
+      const auctionAddress = await riskManagerV1.depositToAuction(
+        tbtcDeposit1.address
+      )
+      expect(auctionAddress).to.equal(ZERO_ADDRESS)
+
+      // 1.5 TBTC should be left on Risk Manager after purchasing 0xfc9..
+      // deposit with surplus funds
+      const tbtcSurplus = await tbtcToken.balanceOf(riskManagerV1.address)
+      expect(tbtcSurplus).to.be.equal(to1ePrecision(15, 17))
```
[the remark this answers is not in the document]
Comment (dimpar) [reply test/system/surplus-for-auction.test.js:157]: Done

File: test/system/surplus-for-auction.test.js (hunk 4 above)
Comment (dimpar) [reply test/system/surplus-for-auction.test.js]: Done

**Review by lukasz-zimnoch — APPROVED** [review]

File: test/system/constants.js [hunk 13]
```diff
@@ -3,8 +3,11 @@ module.exports.keepTokenAddress = "0x85Eee30c52B0b379b046Fb0F85F4f3Dc3009aFEC"
 module.exports.tbtcDepositTokenAddress =
   "0x10b66bd1e3b5a936b7f8dbc5976004311037cdf0"
 // https://allthekeeps.com/deposit/0x55d8b1dd88e60d12c81b5479186c15d07555db9d)
-module.exports.depositAddress = "0x55d8b1dd88e60d12c81b5479186c15d07555db9d"
-module.exports.bidderAddress = "0xa0216ED2202459068a750bDf74063f677613DA34"
+module.exports.depositAddress1 = "0x55d8b1dd88e60d12c81b5479186c15d07555db9d"
+module.exports.depositAddress2 = "0x8495732aecd7f132eaab61f64858ccc73475973f"
+module.exports.depositAddress3 = "0xfc9c50fd44879bd7085edd311bc8e2b7d3e41595"
```
Comment (pdyraga) [inline test/system/constants.js:8]: Would be good to note in a comment what are the differences between those deposits and when I'd choose which one.

Comment (dimpar) [reply test/system/constants.js:8]: Sure, added description here https://github.com/keep-network/coverage-pools/pull/113

**Pull request merged**